Reduce space needed by search and help top menu item

oliverguenther opened the pull request:
This commit changes two top menu items:
1. Search is by default collapsed to its magnifier icon
   - Upon clicking it, it is expanded and focus given to the search
     field
   - Clicking any hierarchy not in below the wrapper, the search is
     collapsed again
2. The help button remains as a '?' icon, and loses the caption.

Relevant work package:
https://community.openproject.org/work_packages/20540

oliverguenther committed:
Reduce space needed by search and help top menu item

This commit changes two top menu items:

1. Search is by default collapsed to its magnifier icon
   - Upon clicking it, it is expanded and focus given to the search
   field
   - Clicking any hierarchy not in below the wrapper, the search is
   collapsed again

2. The help button remains as a '?' icon, and loses the caption.

Relevant work package:
https://community.openproject.org/work_packages/20540

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-reduce-top-menu-space.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-reduce-top-menu-space.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented on config/initializers/menus.rb:
```diff
@@ -42,7 +42,7 @@
   menu.push :help, OpenProject::Info.help_url,
             last: true,
-            caption: I18n.t('label_help'),
+            caption: '',
```
I might be missing something and also did not try it out, but i believe you just killed accessibility:

It seems that there is no text left that a screen reader can use to explain where the help button leads.

oliverguenther replied:
Yeah, that one's not pretty.Currently, caption is used both as the content and the title of the menu item.

I can extend the Redmine menu system to accept a separate title attribute. Would you accept using the html hash for the title ?

i.e.,

```
caption: '',
html: { title: I18n.t('label_help'), accesskey: OpenProject::AccessKeys.key_for(:help), class: 'icon5 icon-help' }
```

NobodysNightmare replied:
Accepting a `title` parameter that defaults to the `caption` sounds pretty clean to me (and should be straight forward using named parameters... probably it is old code and named parameters won't work...).

Using the `html` hash sounds like a reasonable workaround, but TBH I can't tell if this is rather a neat trick or a dirty hack.

oliverguenther replied:
As we only need to set the html title, I argue that allowing an override in the html options is the simplest solution. I have updated the PR to show the simplicity of that solution.

The title is internally set to `html_options[:title]` [(cf. menu_helper.rb)](https://github.com/opf/openproject/pull/3160/files#diff-8c79271c9d2d36c34932fb68d5560378R203), allowing a very  simplest override.

NobodysNightmare commented:
When the search is collapsed it seems to be out of the tab-order, from **Modules** I will tab to the help button and not hit the search field.

NobodysNightmare commented:
I don't know what is intended here, but I noticed a difference between mouse interaction and keyboard interaction and the search field.

**Mouse only**:
- click the magnifier button and the search text pops up
- click outside the textbox and the textbox is hidden again

**Keyboard only**:
- (could not test how to open the textbox using keyboard)
- tabbing out of the textbox does _not_ hide it

NobodysNightmare commented:
Ah, more accessibility issues:

Pressing `s` on the keyboard used to focus the searchbox and allowed me to enter my search term directly, but this is not possible anymore.

Thanks to @RobinWagner for pointing out how to quickly access the search box.

NobodysNightmare commented:
I did what everyone hates: I tried it in IE!

When you enter some text the IE displays a cross to clear the entered text. Normally it looks like this:

![ie-good-cross](https://cloud.githubusercontent.com/assets/453584/8285128/b645da00-1901-11e5-994a-49eaaba2055f.png)

But on your branch it looks like this:

![ie-bad-cross](https://cloud.githubusercontent.com/assets/453584/8285154/e1aff626-1901-11e5-885f-567ea6866608.png)

It is also partially covered by the hitbox of the magnifier, making it hard to click the cross...

NobodysNightmare commented on app/assets/javascripts/top_menu.js:
```diff
@@ -241,8 +241,39 @@
+      , form = wrapper.closest('form')
+      , btn = wrapper.find('.search_btn')
+      , input = wrapper.find('.search_field')
+      , evtns = 'click.expandableSearch';
```
To start with some nitpicking: Did you check that style with the rest of the code? Not that our codebase is particulary homogenous, but I think we usually pute the `,` onto the upper line, without a space before it...

Still this is not written down anywhere I know...

floriank replied:
_loadsShotgun_

:gun:

oliverguenther replied:
This follows NPM's comma-first style, give me a style guide to follow before the shrapnel lands ;)

floriank replied:
I think it was `gulp lint` or jshint that can give you pointers.

And I shalleth not careth what them NPM heathens doeth!

oliverguenther committed:
Fixing Accessibility issues

- Allow s, alt+4 to focus search form
- Collapse search on tab out
- `::-ms-clear` size and position fix

The search form now behaves as following

**Keyboard**

  * Tab until collapsed search (icon) is focused, the search should expand and the input gains focus.
  * Tab out of the input, icon will gain focus, search should NOT collapse.
  * Press enter while icon is focused, search form should be submitted.
  * Tab out of icon to help, search should collapse._
  * Pressing `s` or [modifier + 4](https://en.wikipedia.org/wiki/Access_key) should expand the search.

**Mouse**

  * Click on collapsed magnifier, search should expand.
  * Click on any other element other than input / icon, search should collapse.

**Both**

  * Icon is now flipped when expanded (to resemble the previous state). **Should the collapsed search icon be flipped as well ?**

oliverguenther commented:
Thanks Jan et. al, I owe you a beer once I get to Berlin.
- s, alt+4 to focus search query is restored
- collapse on focusout of any item outside .search_wrapper

oliverguenther commented:
The search icon is not focusable due to the non-obtrusive anchor missing href, at least on some browsers. [Adding tabindex="0" seems like a reasonable solution](https://nemisj.com/focusable/) and should solve the problem.

As for the `::-ms-clear` sizing, that is puzzling - it apparently has no size rule and the outer input size hasn't changed, but I've now forced its width to equal the icon font width without padding.

![bildschirmfoto 2015-06-22 um 19 23 20](https://cloud.githubusercontent.com/assets/459462/8288090/2bc18c54-1914-11e5-8f47-a91039947973.png)

### The search form now behaves as following

(new  highlighted in _italics_)

**Keyboard**
- Tab until collapsed search (icon) is focused, the search should expand and the input gains focus.
- _Tab out of the input, icon will gain focus, search should NOT collapse._
- _Press enter while icon is focused, search form should be submitted._
- Tab out of icon to help, search should collapse._
- Pressing `s` or [modifier + 4](https://en.wikipedia.org/wiki/Access_key) should expand the search.

**Mouse**
- Click on collapsed magnifier, search should expand.
- Click on any other element other than input / icon, search should collapse.
- Click on the magnifier while the search is expanded, search should be submitted.

**Both** 
- Icon is now flipped when expanded (to resemble the previous state). **Should the collapsed search icon be flipped as well ?**

oliverguenther committed:
Allow overriding the html title for accessible help

NobodysNightmare commented:
From a users standpoint I can find no further problems. I did not care very much for the flipped magnifier icon, but the actual behaviour of it changing orientation on collapse/expand reminds me of other toggleable UI elements, so I think I like the current behaviour.

I also like that the new searchbox (opposed to the state before) now has a distinct submit button that I can reach via tab, I am no accessibility expert by any means, but I think this is an improvement.

I'll have another look at the code while travis is deciding whether the tests passed or not, but I plan on merging this before lunch.

P.S.: Just noticed that there is still a WIP label, I'll only merge if you remove that or indicate in any other way that you feel you are done with your work ;-)

myabc commented:
@oliverguenther FYI, this doesn't really fit in with the Sass code style I tried to establish. I ran out of time to document it in the style guide itself, but it broadly follows what's described in [this article](http://www.hagenburger.net/BLOG/Modular-CSS-Class-Names.html). I'd be happy to elaborate though, if that's helpful.

oliverguenther commented:
Thanks again Jan for your thorough review. Please hold back until I had another look at the Sass style guide from Alex.

NobodysNightmare commented on app/assets/stylesheets/layout/_top_menu.sass:
```diff
@@ -51,37 +51,51 @@
-    top:   9px
-    right: 9px
-    &:after
+    display: none
```
I actually liked the old rule (`display: none` for `.search-collapsed .search_field`) more. Because there is no apparent reason why a lonely `search_field` should not be displayed, for a collapsed search field this is more obvious.

Was there a technical reason for that change in https://github.com/oliverguenther/openproject/commit/3b2027b42cf3956e4803a9295347b5d8dab4cdb6?

oliverguenther replied:
The initial PR had a two-class state (search-collapsed, search-expanded) for actually handling a single-bit state, thus I changed it into no class = collapsed, expanded = expanded. 

We could turn this state around, such that we re-introduce `search-collapsed` which is removed upon expanding.

NobodysNightmare replied:
Ah, I see. That makes sense. In this case I agree that the inversion would be the right thing to do, so that without any modifiers the search is visible.

Also after reading the rules provided by @myabc it should probably be `-collapsed` instead of `search-collapsed`, which seems to make sense...

myabc commented on app/assets/javascripts/top_menu.js:
```diff
@@ -241,8 +241,48 @@
+
+  // expands .search_btn inside the given wrapper
+  // to an input .search_field with the button appended to the right.
+  $.fn.expandableSearch = function () {
```
@oliverguenther why don't you use an Angular directive?

oliverguenther committed:
Rewrite search box as angular directive

oliverguenther commented on frontend/app/ui_components/expandable-search.js:
```diff
@@ -0,0 +1,73 @@
+          // Submit only when clicked or enter on btn
+          if (evt.type === 'mousedown' ||
+             (evt.type === 'keypress' && evt.which === ENTER_KEY)) {
+            elem.closest('form').submit();
```
I've rewritten the  search functionality as a directive, but can't seem to find the correct way to locate the parent (angular) form surrounding the search input. `input.form` is `undefined`, and this workaround doesn't feel right. Can you point me to the right direction?

NobodysNightmare replied:
I can't help you on that one either. Maybe @myabc has any valuable input?

I can think of two ways out:
1. the one you have chosen: The form is not part of the directive, but you expect the directive to be used within a form (still don't know if `elem.closest` is the best possible solution here)
2. What I could also imagine is to include the form **in** your directive. After all you need that form, so it might be a good idea to make it a part of your component. This should ease finding it, but might introduce other problems I did not yet see.

I would merge it either way, but I am also happy to receive further feedback beforehand.

@oliverguenther Tell me when you feel ready enough ;-)

oliverguenther replied:
Thanks Jan, If possible I would like to avoid moving form inside the directive, since it originates from erb and I'm not sure how much effort including it in the directive would be. I'll keep the WIP tag to give Alex had the chance to comment.

@myabc Do you have concerns regarding the `closest` and can point me to a better solution?

oliverguenther commented on app/assets/stylesheets/layout/_top_menu.sass:
```diff
@@ -181,6 +154,55 @@
+  &:hover
+    text-decoration: none
+
+  .top-menu-search:not(.-collapsed) &
```
I agree with and am following the sass style guide Alex linked before.
For the single-state toggle, assuming expanded is the default rule forces me to either use this inverted rule or override several values in the collapsed state (which I would rather avoid). Do you have any concern regarding the `:not` rule?

oliverguenther replied:
I agree and will follow the style guide Alex linked before.
For the single-state toggle, assuming expanded is the default rule forces me to either use this style or override several values in the collapsed state (which I would rather avoid). Do you have any concern regarding the `:not` rule?

NobodysNightmare replied:
Using `:not` should be okay IMHO

ulferts commented:
Spec failure was due to a poorly written spec not related to this PR. I restarted the spec run.

oliverguenther commented:
@NobodysNightmare Ready to merge from my side.
